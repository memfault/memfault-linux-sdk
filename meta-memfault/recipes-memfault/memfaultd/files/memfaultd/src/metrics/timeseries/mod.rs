//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use chrono::{DateTime, Utc};
use eyre::{eyre, Result};
use std::cmp;

use super::{construct_histogram_value, MetricReading, MetricValue};

const FINITENESS_ERROR: &str = "Metric values must be finite.";

/// A trait for the storage of multiple metric events aggregated together.
/// This (roughly) maps to a time series in OpenTelemetry data model:
/// https://opentelemetry.io/docs/specs/otel/metrics/data-model/#timeseries-model
pub trait TimeSeries {
    fn aggregate(&mut self, newer: &MetricReading) -> Result<()>;
    fn value(&self) -> MetricValue;
}

pub struct Histogram {
    sum: f64,
    count: u64,
    start: DateTime<Utc>,
    end: DateTime<Utc>,
    min: f64,
    max: f64,
}

impl Histogram {
    pub fn new(reading: &MetricReading) -> Result<Self> {
        match *reading {
            MetricReading::Histogram { value, timestamp } => {
                if !value.is_finite() {
                    return Err(eyre!(FINITENESS_ERROR));
                }

                Ok(Self {
                    sum: value,
                    count: 1,
                    start: timestamp,
                    end: timestamp,
                    min: value,
                    max: value,
                })
            }
            _ => Err(eyre!("Cannot create a histogram from a non-gauge metric")),
        }
    }
}

impl TimeSeries for Histogram {
    fn aggregate(&mut self, newer: &MetricReading) -> Result<()> {
        match newer {
            MetricReading::Histogram { value, timestamp } => {
                if !value.is_finite() {
                    return Err(eyre!(FINITENESS_ERROR));
                }
                self.sum += value;
                self.count += 1;
                self.start = cmp::min(self.start, *timestamp);
                self.end = cmp::max(self.end, *timestamp);
                self.min = f64::min(self.min, *value);
                self.max = f64::max(self.max, *value);
                Ok(())
            }
            _ => Err(eyre!(
                "Cannot aggregate a histogram with a non-gauge metric"
            )),
        }
    }

    fn value(&self) -> MetricValue {
        if self.count > 0 {
            MetricValue::Histogram(construct_histogram_value(
                self.min,
                self.sum / self.count as f64,
                self.max,
            ))
        } else {
            MetricValue::Histogram(construct_histogram_value(f64::NAN, f64::NAN, f64::NAN))
        }
    }
}

/// An aggregation that calculates the sum of all values received. This assumes that all readings will be positive numbers.
/// Monotonic counter in OpenTelemetry data model.
pub struct Counter {
    sum: f64,
    end: DateTime<Utc>,
}

impl Counter {
    pub fn new(reading: &MetricReading) -> Result<Self> {
        match *reading {
            MetricReading::Counter { value, timestamp } => {
                if !value.is_finite() {
                    return Err(eyre!(FINITENESS_ERROR));
                }
                Ok(Self {
                    sum: value,
                    end: timestamp,
                })
            }
            _ => Err(eyre!("Cannot create a sum from a non-counter metric")),
        }
    }
}

impl TimeSeries for Counter {
    fn aggregate(&mut self, newer: &MetricReading) -> Result<()> {
        match newer {
            MetricReading::Counter { value, timestamp } => {
                if !value.is_finite() {
                    return Err(eyre!(FINITENESS_ERROR));
                }
                self.sum += value;
                self.end = *timestamp;
                Ok(())
            }
            _ => Err(eyre!("Cannot aggregate a sum with a non-counter metric")),
        }
    }

    fn value(&self) -> MetricValue {
        MetricValue::Number(self.sum)
    }
}

/// A time-weighted sum of all values received. This is useful to maintain an accurate average measurement when the interval between readings is not constant.
pub struct TimeWeightedAverage {
    weighted_sum: f64,
    duration: u64,
    end: DateTime<Utc>,
}

impl TimeWeightedAverage {
    pub fn new(reading: &MetricReading) -> Result<Self> {
        match *reading {
            MetricReading::TimeWeightedAverage {
                value,
                timestamp,
                interval,
            } => {
                if !value.is_finite() {
                    return Err(eyre!(FINITENESS_ERROR));
                }
                Ok(Self {
                    weighted_sum: value * interval.num_milliseconds() as f64,
                    duration: interval.num_milliseconds() as u64,
                    end: timestamp,
                })
            }
            _ => Err(eyre!(
                "Mismatch between Time Weighted Average aggregation and {:?} reading",
                reading
            )),
        }
    }
}

impl TimeSeries for TimeWeightedAverage {
    fn aggregate(&mut self, newer: &MetricReading) -> Result<()> {
        match newer {
            MetricReading::TimeWeightedAverage {
                value, timestamp, ..
            } => {
                if !value.is_finite() {
                    return Err(eyre!(FINITENESS_ERROR));
                }
                if timestamp < &self.end {
                    return Err(eyre!(
                        "Cannot aggregate a time-weighted average with an older timestamp"
                    ));
                }
                let duration = (*timestamp - self.end).num_milliseconds() as u64;
                self.weighted_sum += value * duration as f64;
                self.duration += duration;
                self.end = *timestamp;
                Ok(())
            }
            _ => Err(eyre!(
                "Cannot aggregate a time-weighted average with a non-gauge metric"
            )),
        }
    }

    fn value(&self) -> MetricValue {
        if self.duration > 0 {
            MetricValue::Number(self.weighted_sum / self.duration as f64)
        } else {
            MetricValue::Number(f64::NAN)
        }
    }
}

pub struct Gauge {
    value: f64,
    end: DateTime<Utc>,
}

impl Gauge {
    pub fn new(reading: &MetricReading) -> Result<Self> {
        match *reading {
            MetricReading::Gauge { value, timestamp } => {
                if !value.is_finite() {
                    return Err(eyre!(FINITENESS_ERROR));
                }
                Ok(Self {
                    value,
                    end: timestamp,
                })
            }
            _ => Err(eyre!(
                "Cannot create a gauge aggregation from a non-gauge metric"
            )),
        }
    }
}

impl TimeSeries for Gauge {
    fn aggregate(&mut self, newer: &MetricReading) -> Result<()> {
        match newer {
            MetricReading::Gauge { value, timestamp } => {
                if !value.is_finite() {
                    return Err(eyre!(FINITENESS_ERROR));
                }
                if *timestamp > self.end {
                    self.value = *value;
                    self.end = *timestamp;
                }
                Ok(())
            }
            _ => Err(eyre!(
                "Cannot aggregate a histogram with a non-gauge metric"
            )),
        }
    }

    fn value(&self) -> MetricValue {
        MetricValue::Number(self.value)
    }
}

/// An aggregation stores the most recently received String
/// associated with a key as a tag on the report
pub struct ReportTag {
    value: String,
    end: DateTime<Utc>,
}

impl ReportTag {
    pub fn new(reading: &MetricReading) -> Result<Self> {
        match reading {
            MetricReading::ReportTag { value, timestamp } => Ok(Self {
                value: value.clone(),
                end: *timestamp,
            }),
            _ => Err(eyre!(
                "Cannot create a report tag from a non-report tag reading"
            )),
        }
    }
}

impl TimeSeries for ReportTag {
    fn aggregate(&mut self, newer: &MetricReading) -> Result<()> {
        match newer {
            MetricReading::ReportTag { value, timestamp } => {
                if *timestamp > self.end {
                    self.value = value.clone();
                    self.end = *timestamp;
                }
                Ok(())
            }
            _ => Err(eyre!(
                "Cannot aggregate a report tag with a non-report-tag reading"
            )),
        }
    }

    fn value(&self) -> MetricValue {
        MetricValue::String(self.value.clone())
    }
}
#[cfg(test)]
mod tests {
    use chrono::Duration;
    use rstest::rstest;

    use crate::metrics::{HistogramValue, MetricReading, MetricTimestamp, MetricValue};
    use std::{f64::INFINITY, f64::NAN, f64::NEG_INFINITY, str::FromStr};

    use super::TimeSeries;
    use super::{Counter, Gauge, Histogram, TimeWeightedAverage};

    #[rstest]
    #[case(1.0, 1000, 2.0, 1.5, 2.0, 1.0, 1000)]
    #[case(10.0, 10_000, 10.0, 10.0, 10.0, 10.0, 10_000)]
    #[case(1.0, 9_000, 0.0, 0.5, 1.0, 0.0, 9_000)]
    #[case(1.0, 0, 2.0, 1.5, 2.0, 1.0, 0)]
    fn test_histogram_aggregation(
        #[case] a: f64,
        #[case] duration_between_ms: i64,
        #[case] b: f64,
        #[case] expected_mean: f64,
        #[case] expected_max: f64,
        #[case] expected_min: f64,
        #[case] expected_ms: i64,
    ) {
        let t0 = MetricTimestamp::from_str("2021-01-01T00:00:00Z").unwrap();

        let a = MetricReading::Histogram {
            value: a,
            timestamp: t0,
        };
        let b = MetricReading::Histogram {
            value: b,
            timestamp: t0 + Duration::milliseconds(duration_between_ms),
        };

        let mut h = Histogram::new(&a).unwrap();
        h.aggregate(&b).unwrap();

        assert_eq!(h.start, t0);
        assert_eq!(h.end, t0 + Duration::milliseconds(duration_between_ms));
        assert_eq!((h.end - h.start).num_milliseconds(), expected_ms);
        assert_eq!(
            h.value(),
            MetricValue::Histogram(HistogramValue {
                min: expected_min,
                mean: expected_mean,
                max: expected_max,
            })
        );
    }

    #[rstest]
    #[case(1.0, 1000, 2.0, 1000, 1.5, 2000)]
    #[case(10.0, 10000, 10.0, 1000, 10.0, 11000)]
    #[case(1.0, 9_000, 0.0, 1_000, 0.9, 10_000)]
    #[case(1.0, 0, 2.0, 1, 2.0, 1)]
    #[case(1.0, 1000, 2.0, 0, 1.0, 1000)]
    fn test_time_weighted_aggregation(
        #[case] a: f64,
        #[case] a_ms: i64,
        #[case] b: f64,
        #[case] b_ms: i64,
        #[case] expected: f64,
        #[case] expected_ms: u64,
    ) {
        let t0 = MetricTimestamp::from_str("2021-01-01T00:00:00Z").unwrap();

        let a = MetricReading::TimeWeightedAverage {
            value: a,
            timestamp: t0 + Duration::milliseconds(a_ms),
            interval: Duration::milliseconds(a_ms),
        };
        let b = MetricReading::TimeWeightedAverage {
            value: b,
            timestamp: t0 + Duration::milliseconds(a_ms + b_ms),
            interval: Duration::milliseconds(b_ms),
        };

        let mut h = TimeWeightedAverage::new(&a).unwrap();
        h.aggregate(&b).unwrap();

        assert_eq!(h.end, t0 + Duration::milliseconds(a_ms + b_ms));
        assert_eq!(h.duration, expected_ms);
        assert_eq!(h.value(), MetricValue::Number(expected));
    }

    #[rstest]
    fn test_incompatible_metric_type_on_histogram() {
        let timestamp = MetricTimestamp::from_str("2021-01-01T00:00:00Z").unwrap();

        let a = MetricReading::Histogram {
            value: 1.0,
            timestamp,
        };
        let b = MetricReading::Counter {
            value: 2.0,
            timestamp,
        };

        assert!(Histogram::new(&a).unwrap().aggregate(&b).is_err());
    }

    #[rstest]
    #[case(INFINITY)]
    #[case(NEG_INFINITY)]
    #[case(NAN)]
    fn test_edge_values_new(#[case] edge_value: f64) {
        let timestamp = MetricTimestamp::from_str("2021-01-01T00:00:00Z").unwrap();
        let a = MetricReading::Histogram {
            value: edge_value,
            timestamp,
        };
        assert!(Histogram::new(&a).is_err());
    }

    #[rstest]
    #[case(INFINITY)]
    #[case(NEG_INFINITY)]
    #[case(NAN)]
    fn test_edge_values_aggregate(#[case] edge_value: f64) {
        let timestamp = MetricTimestamp::from_str("2021-01-01T00:00:00Z").unwrap();
        let a = MetricReading::Histogram {
            value: 0.0,
            timestamp,
        };
        let b = MetricReading::Histogram {
            value: edge_value,
            timestamp,
        };
        assert!(Histogram::new(&a).unwrap().aggregate(&b).is_err());
    }

    #[rstest]
    #[case(1.0, 2.0, 3.0)]
    fn test_counter_aggregation(#[case] a: f64, #[case] b: f64, #[case] expected: f64) {
        let timestamp = MetricTimestamp::from_str("2021-01-01T00:00:00Z").unwrap();
        let timestamp2 = MetricTimestamp::from_str("2021-01-01T00:00:43Z").unwrap();

        let a = MetricReading::Counter {
            value: a,
            timestamp,
        };
        let b = MetricReading::Counter {
            value: b,
            timestamp: timestamp2,
        };

        let mut sum = Counter::new(&a).unwrap();
        sum.aggregate(&b).unwrap();
        assert_eq!(sum.end, timestamp2);
        assert_eq!(sum.sum, expected);
    }

    #[rstest]
    #[case(1.0, 2.0, 2.0)]
    #[case(10.0, 1.0, 1.0)]
    fn test_gauge_aggregation(#[case] a: f64, #[case] b: f64, #[case] expected: f64) {
        let timestamp = MetricTimestamp::from_str("2021-01-01T00:00:00Z").unwrap();
        let timestamp2 = MetricTimestamp::from_str("2021-01-01T00:00:43Z").unwrap();

        let a = MetricReading::Gauge {
            value: a,
            timestamp,
        };
        let b = MetricReading::Gauge {
            value: b,
            timestamp: timestamp2,
        };

        let mut gauge = Gauge::new(&a).unwrap();
        gauge.aggregate(&b).unwrap();
        assert_eq!(gauge.end, timestamp2);
        assert_eq!(gauge.value, expected);
    }
}
