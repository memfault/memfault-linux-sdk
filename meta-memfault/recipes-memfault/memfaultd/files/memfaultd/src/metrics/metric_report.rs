//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use chrono::Utc;
use eyre::{eyre, Result};
use serde::{Deserialize, Serialize};
use std::{
    collections::{HashMap, HashSet},
    path::Path,
    str::FromStr,
    time::{Duration, Instant},
};

use crate::{
    mar::{MarEntryBuilder, Metadata},
    metrics::{
        core_metrics::{
            METRIC_BATTERY_DISCHARGE_DURATION_MS, METRIC_BATTERY_SOC_PCT_DROP,
            METRIC_CONNECTED_TIME, METRIC_EXPECTED_CONNECTED_TIME, METRIC_MF_SYNC_FAILURE,
            METRIC_MF_SYNC_SUCCESS, METRIC_OPERATIONAL_CRASHES, METRIC_SYNC_FAILURE,
            METRIC_SYNC_SUCCESS,
        },
        metric_reading::KeyedMetricReading,
        timeseries::{Counter, Gauge, Histogram, TimeSeries, TimeWeightedAverage},
        MetricReading, MetricStringKey, MetricValue, SessionName,
    },
};

use super::timeseries::ReportTag;

pub enum CapturedMetrics {
    All,
    Metrics(HashSet<MetricStringKey>),
}

pub const HEARTBEAT_REPORT_TYPE: &str = "heartbeat";
pub const DAILY_HEARTBEAT_REPORT_TYPE: &str = "daily-heartbeat";

const SESSION_CORE_METRICS: &[&str; 9] = &[
    METRIC_MF_SYNC_FAILURE,
    METRIC_MF_SYNC_SUCCESS,
    METRIC_BATTERY_DISCHARGE_DURATION_MS,
    METRIC_BATTERY_SOC_PCT_DROP,
    METRIC_CONNECTED_TIME,
    METRIC_EXPECTED_CONNECTED_TIME,
    METRIC_SYNC_FAILURE,
    METRIC_SYNC_SUCCESS,
    METRIC_OPERATIONAL_CRASHES,
];

#[derive(Serialize, Deserialize, Clone, PartialEq, Eq)]
pub enum MetricReportType {
    #[serde(rename = "heartbeat")]
    Heartbeat,
    #[serde(rename = "session")]
    Session(SessionName),
    #[serde(rename = "daily-heartbeat")]
    DailyHeartbeat,
}

impl MetricReportType {
    pub fn as_str(&self) -> &str {
        match self {
            MetricReportType::Heartbeat => HEARTBEAT_REPORT_TYPE,
            MetricReportType::Session(session_name) => session_name.as_str(),
            MetricReportType::DailyHeartbeat => DAILY_HEARTBEAT_REPORT_TYPE,
        }
    }
}

pub struct MetricReport {
    /// In-memory metric store for this report
    metrics: HashMap<MetricStringKey, Box<dyn TimeSeries + Send>>,
    /// Point in time when capture of metrics currently in
    /// report's metric store began
    start: Instant,
    /// Configuration of which metrics this report should capture
    captured_metrics: CapturedMetrics,
    /// Indicates whether this is a heartbeat metric report or
    /// session metric report (with session name)
    report_type: MetricReportType,
}

struct MetricReportSnapshot {
    duration: Duration,
    metrics: HashMap<MetricStringKey, MetricValue>,
}

impl MetricReport {
    pub fn new(
        report_type: MetricReportType,
        configured_captured_metrics: CapturedMetrics,
    ) -> Self {
        // Always include session core metrics regardless of configuration
        let captured_metrics = match configured_captured_metrics {
            CapturedMetrics::All => CapturedMetrics::All,
            CapturedMetrics::Metrics(metrics) => {
                let mut merged_set = SESSION_CORE_METRICS
                    .iter()
                    .map(|core_metric_key| {
                        // This expect should never be hit as SESSION_CORE_METRICS is
                        // an array of static strings
                        MetricStringKey::from_str(core_metric_key)
                            .expect("Invalid Metric Key in SESSION_CORE_METRICS")
                    })
                    .collect::<HashSet<_>>();
                merged_set.extend(metrics);
                CapturedMetrics::Metrics(merged_set)
            }
        };

        Self {
            metrics: HashMap::new(),
            start: Instant::now(),
            captured_metrics,
            report_type,
        }
    }

    /// Creates a heartbeat report that captures all metrics
    pub fn new_heartbeat() -> Self {
        MetricReport::new(MetricReportType::Heartbeat, CapturedMetrics::All)
    }

    /// Creates a daily heartbeat report that captures all metrics
    pub fn new_daily_heartbeat() -> Self {
        MetricReport::new(MetricReportType::DailyHeartbeat, CapturedMetrics::All)
    }

    fn is_captured(&self, metric_key: &MetricStringKey) -> bool {
        match &self.captured_metrics {
            CapturedMetrics::Metrics(metric_keys) => metric_keys.contains(metric_key),
            CapturedMetrics::All => true,
        }
    }

    /// Adds a metric reading to the report's internal
    /// metric store if the report captures that metric,
    /// otherwise no-op
    pub fn add_metric(&mut self, m: KeyedMetricReading) -> Result<()> {
        if self.is_captured(&m.name) {
            match self.metrics.entry(m.name) {
                std::collections::hash_map::Entry::Occupied(mut o) => {
                    let state = o.get_mut();
                    if let Err(e) = (*state).aggregate(&m.value) {
                        *state = Self::select_aggregate_for(&m.value)?;
                        log::warn!(
                            "New value for metric {} is incompatible ({}). Resetting timeseries.",
                            o.key(),
                            e
                        );
                    }
                }
                std::collections::hash_map::Entry::Vacant(v) => {
                    let timeseries = Self::select_aggregate_for(&m.value)?;
                    v.insert(timeseries);
                }
            };
        }
        Ok(())
    }

    /// Increment a counter metric by 1
    pub fn increment_counter(&mut self, name: &str) -> Result<()> {
        self.add_to_counter(name, 1.0)
    }

    pub fn add_to_counter(&mut self, name: &str, value: f64) -> Result<()> {
        match name.parse::<MetricStringKey>() {
            Ok(metric_name) => self.add_metric(KeyedMetricReading::new(
                metric_name,
                MetricReading::Counter {
                    value,
                    timestamp: Utc::now(),
                },
            )),
            Err(e) => Err(eyre!("Invalid metric name: {} - {}", name, e)),
        }
    }

    /// Return all the metrics in memory for this report and resets its store.
    pub fn take_metrics(&mut self) -> HashMap<MetricStringKey, MetricValue> {
        self.take_metric_report_snapshot().metrics
    }

    fn take_metric_report_snapshot(&mut self) -> MetricReportSnapshot {
        let duration = std::mem::replace(&mut self.start, Instant::now()).elapsed();
        let metrics = std::mem::take(&mut self.metrics)
            .into_iter()
            .map(|(name, state)| (name, state.value()))
            .collect();

        MetricReportSnapshot { duration, metrics }
    }

    /// Create one metric report MAR entry with all the metrics in the store.
    ///
    /// All data will be timestamped with current time measured by CollectionTime::now(), effectively
    /// disregarding the collectd timestamps.
    pub fn prepare_metric_report(
        &mut self,
        mar_staging_area: &Path,
    ) -> Result<Option<MarEntryBuilder<Metadata>>> {
        let snapshot = self.take_metric_report_snapshot();

        if snapshot.metrics.is_empty() {
            return Ok(None);
        }

        Ok(Some(MarEntryBuilder::new(mar_staging_area)?.set_metadata(
            Metadata::new_metric_report(
                snapshot.metrics,
                snapshot.duration,
                self.report_type.clone(),
            ),
        )))
    }

    fn select_aggregate_for(event: &MetricReading) -> Result<Box<dyn TimeSeries + Send>> {
        match event {
            MetricReading::Histogram { .. } => Ok(Box::new(Histogram::new(event)?)),
            MetricReading::Counter { .. } => Ok(Box::new(Counter::new(event)?)),
            MetricReading::Gauge { .. } => Ok(Box::new(Gauge::new(event)?)),
            MetricReading::TimeWeightedAverage { .. } => {
                Ok(Box::new(TimeWeightedAverage::new(event)?))
            }
            MetricReading::ReportTag { .. } => Ok(Box::new(ReportTag::new(event)?)),
        }
    }

    pub fn report_type(&self) -> &MetricReportType {
        &self.report_type
    }
}

#[cfg(test)]
mod tests {
    use tempfile::TempDir;

    use std::collections::BTreeMap;

    use super::*;
    use crate::test_utils::in_histograms;
    use std::str::FromStr;

    use insta::assert_json_snapshot;
    use rstest::rstest;

    #[rstest]
    #[case(in_histograms(vec![("foo", 1.0), ("bar", 2.0), ("baz",  3.0)]), "heartbeat_report_1")]
    #[case(in_histograms(vec![("foo", 1.0), ("foo", 2.0), ("foo",  3.0)]), "heartbeat_report_2")]
    #[case(in_histograms(vec![("foo", 1.0), ("foo", 1.0)]), "heartbeat_report_3")]
    #[case(in_histograms(vec![("foo", 1.0), ("foo", 2.0)]), "heartbeat_report_4")]
    #[case(in_histograms(vec![("foo", 1.0), ("foo", 2.0), ("foo",  2.0)]), "heartbeat_report_5")]
    fn test_aggregate_metrics(
        #[case] metrics: impl Iterator<Item = KeyedMetricReading>,
        #[case] test_name: &str,
    ) {
        let mut metric_report = MetricReport::new_heartbeat();

        for m in metrics {
            metric_report.add_metric(m).unwrap();
        }
        let sorted_metrics: BTreeMap<_, _> = metric_report.take_metrics().into_iter().collect();
        assert_json_snapshot!(test_name, sorted_metrics);
    }

    #[rstest]
    #[case(in_histograms(vec![("foo", 1.0), ("bar", 2.0), ("baz",  3.0)]), "session_report_1")]
    #[case(in_histograms(vec![("foo", 1.0), ("foo", 2.0), ("foo",  3.0)]), "session_report_2")]
    #[case(in_histograms(vec![("foo", 1.0), ("foo", 1.0)]), "session_report_3")]
    #[case(in_histograms(vec![("foo", 1.0), ("foo", 2.0)]), "session_report_4")]
    #[case(in_histograms(vec![("foo", 1.0), ("foo", 2.0), ("baz",  2.0), ("bar",  3.0)]), "session_report_5")]
    fn test_aggregate_metrics_session(
        #[case] metrics: impl Iterator<Item = KeyedMetricReading>,
        #[case] test_name: &str,
    ) {
        let mut metric_report = MetricReport::new(
            MetricReportType::Session(SessionName::from_str("foo_only").unwrap()),
            CapturedMetrics::Metrics(
                [
                    MetricStringKey::from_str("foo").unwrap(),
                    MetricStringKey::from_str("baz").unwrap(),
                ]
                .into_iter()
                .collect(),
            ),
        );

        for m in metrics {
            metric_report.add_metric(m).unwrap();
        }
        let sorted_metrics: BTreeMap<_, _> = metric_report.take_metrics().into_iter().collect();
        assert_json_snapshot!(test_name, sorted_metrics);
    }

    /// Core metrics should always be captured by sessions even if they are not
    /// configured to do so
    #[rstest]
    fn test_aggregate_core_metrics_session() {
        let mut metric_report = MetricReport::new(
            MetricReportType::Session(SessionName::from_str("foo_only").unwrap()),
            CapturedMetrics::Metrics(
                [
                    MetricStringKey::from_str("foo").unwrap(),
                    MetricStringKey::from_str("baz").unwrap(),
                ]
                .into_iter()
                .collect(),
            ),
        );

        let metrics = in_histograms(
            SESSION_CORE_METRICS
                .map(|metric_name: &'static str| (metric_name, 100.0))
                .to_vec(),
        );

        for m in metrics {
            metric_report.add_metric(m).unwrap();
        }
        let sorted_metrics: BTreeMap<_, _> = metric_report.take_metrics().into_iter().collect();
        assert_json_snapshot!(sorted_metrics);
    }

    #[rstest]
    fn test_empty_after_write() {
        let mut metric_report = MetricReport::new_heartbeat();
        for m in in_histograms(vec![("foo", 1.0), ("bar", 2.0), ("baz", 3.0)]) {
            metric_report.add_metric(m).unwrap();
        }

        let tempdir = TempDir::new().unwrap();
        let _ = metric_report.prepare_metric_report(tempdir.path());
        assert_eq!(metric_report.take_metrics().len(), 0);
    }
}
