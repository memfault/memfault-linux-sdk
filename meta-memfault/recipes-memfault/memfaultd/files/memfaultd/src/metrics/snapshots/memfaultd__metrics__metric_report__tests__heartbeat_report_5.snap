---
source: memfaultd/src/metrics/metric_report.rs
expression: sorted_metrics
---
{
  "memory_systemd_pct": 1.6666666666666667,
  "memory_systemd_pct_max": 2.0,
  "memory_systemd_pct_min": 1.0
}
